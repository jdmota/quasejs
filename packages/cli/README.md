## About

It's like `meow` but includes `update-notifier`.

We also change the update message if Yarn is detected.

## Usage example

`bin/my-name`

```js
#!/usr/bin/env node
"use strict";

require( "@quase/cli" ).default(
  ( { input, flags, pkg, help, showHelp } ) => {

  },
  {}, // Meow options
  {
    options: {}, // UpdateNotifier options
    notify: {} // .notify() options
  }, // false to disable notification
);
```

`package.json`

```json
{
  "bin": {
    "my-name": "bin/my-name"
  },
}
```

See https://github.com/sindresorhus/meow for details.

See https://github.com/yeoman/update-notifier for details.
