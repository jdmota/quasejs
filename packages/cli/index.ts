import { Command, Option } from "clipanion";

// TODO use schema
// TODO use https://github.com/sindresorhus/import-local
// TODO use https://github.com/sindresorhus/update-notifier

export class HelloCommand extends Command {
  name = Option.String();

  async execute() {
    this.context.stdout.write(`Hello ${this.name}!\n`);
  }
}
