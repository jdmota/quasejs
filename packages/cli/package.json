{
  "name": "@quase/cli",
  "version": "0.4.0-0",
  "description": "A simple cli helper.",
  "license": "MIT",
  "homepage": "https://github.com/quase/quasejs",
  "repository": "https://github.com/quase/quasejs/tree/master/packages/cli",
  "module": "src/index.js",
  "main": "dist/index.js",
  "dependencies": {
    "@quase/config": "0.2.0-0",
    "camelcase": "^5.0.0",
    "camelcase-keys": "^4.2.0",
    "currently-unhandled": "^0.4.1",
    "decamelize": "^2.0.0",
    "has-yarn": "^1.0.0",
    "import-local": "^1.0.0",
    "is-installed-globally": "^0.1.0",
    "normalize-package-data": "^2.4.0",
    "read-pkg-up": "^4.0.0",
    "redent": "^2.0.0",
    "signal-exit": "^3.0.2",
    "trim-newlines": "^2.0.0",
    "turbocolor": "^2.4.0",
    "yargs-parser": "^10.0.0"
  },
  "devDependencies": {
    "boxen": "^1.3.0",
    "update-notifier": "^2.5.0"
  }
}
