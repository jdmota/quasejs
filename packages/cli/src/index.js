const chalk = require( "chalk" );
const hasYarn = require( "has-yarn" );
const updateNotifier = require( "update-notifier" );
const meow = require( "meow" );

function notifyFix( opts ) {
  if ( !process.stdout.isTTY || !this.update ) {
    return this;
  }

  opts = Object.assign( { isGlobal: require( "is-installed-globally" ) }, opts );

  const defaultMsg = hasYarn() ?
    `Update available ${chalk.dim( this.update.current )}${chalk.reset( " → " )}${chalk.green( this.update.latest )}` +
    ` \nRun ${chalk.cyan( `yarn ${opts.isGlobal ? "global " : ""}add ${this.packageName}` )} to update` :
    `Update available ${chalk.dim( this.update.current )}${chalk.reset( " → " )}${chalk.green( this.update.latest )}` +
    ` \nRun ${chalk.cyan( `npm i ${opts.isGlobal ? "-g " : ""}${this.packageName}` )} to update`;

  opts.message = opts.message || defaultMsg;

  opts.boxenOpts = opts.boxenOpts || {
    padding: 1,
    margin: 1,
    align: "center",
    borderColor: "yellow",
    borderStyle: "round"
  };

  const message = "\n" + require( "boxen" )( opts.message, opts.boxenOpts );

  /* eslint-disable */
  if ( opts.defer === false ) {
    console.error( message );
  } else {
    process.on( "exit", () => {
      console.error( message );
    } );

    process.on( "SIGINT", () => {
      console.error( "" );
      process.exit();
    } );
  }
  /* eslint-enable */

  return this;
}

function notify( pkg, notifierOpts ) {
  const notifier = updateNotifier( Object.assign( { pkg }, notifierOpts.options ) );
  notifier.notify = notifyFix;
  notifier.notify( notifierOpts.notify );
}

export default function( callback, opts, notifierOpts ) {

  const defaultConfigFile = opts.defaultConfigFile;

  if ( defaultConfigFile ) {
    opts = Object.assign( {}, opts );
    opts.flags = Object.assign( {}, opts.flags );
    opts.flags.config = {
      type: "string",
      alias: "c",
      default: defaultConfigFile
    };
  }

  const cli = meow( opts );

  if ( notifierOpts !== false ) {
    notify( cli.pkg, notifierOpts || {} );
  }

  if ( defaultConfigFile ) {
    if ( cli.flags.config === "none" ) {
      cli.config = {};
    } else {
      cli.config = require( require( "path" ).resolve( cli.flags.config ) );
    }
  }

  callback( cli );
}
