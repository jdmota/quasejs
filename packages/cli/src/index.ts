import hardRejection from "./hard-rejection";
import notify from "./notify";
import { generateHelp as _generateHelp } from "./help";
import { handleArgs } from "./args";
import normalizePkg from "./normalize-pkg";
import { CliOptions } from "./types";
import path from "path";
import { getConfig } from "@quase/config";
import readPkgUp from "read-pkg-up";

const importLocal = require( "import-local" );

/* eslint no-process-exit: 0 */
/* eslint no-console: 0 */

// Prevent caching of this module so module.parent is always accurate
delete require.cache[ __filename ];
const filename = module.parent && module.parent.filename;
const parentDir = ( filename && path.dirname( filename ) ) || undefined;

async function cli( _opts: any ) {
  if ( filename && importLocal( filename ) ) {
    return;
  }
  hardRejection();

  const opts = Object.assign( {
    cwd: process.cwd(),
    inferType: false,
    autoHelp: true,
    autoVersion: true,
    argv: process.argv.slice( 2 )
  }, _opts ) as CliOptions;

  opts.cwd = path.resolve( opts.cwd );

  const pkgJob = opts.pkg ? Promise.resolve( { package: opts.pkg } ) : readPkgUp( {
    cwd: parentDir,
    normalize: false
  } );

  const argsInfo = handleArgs( opts );

  const _pkg = await pkgJob;
  const pkg = normalizePkg( _pkg ? _pkg.package : {} );

  process.title = pkg.bin ? Object.keys( pkg.bin )[ 0 ] : pkg.name;

  if ( !opts.description && opts.description !== false ) {
    opts.description = pkg.description;
  }

  const generateHelp = () => _generateHelp( argsInfo );

  const showHelp = ( code?: number ) => {
    console.log( generateHelp() );
    process.exit( typeof code === "number" ? code : 2 );
  };

  const showVersion = () => {
    console.log( typeof opts.version === "string" ? opts.version : pkg.version );
    process.exit();
  };

  const { schema, commandSet, input, flags } = argsInfo;

  if ( argsInfo.argv.length === 1 ) {
    if ( flags.version === true && opts.autoVersion ) {
      showVersion();
    }

    if ( flags.help === true && opts.autoHelp ) {
      showHelp( 0 );
    }
  }

  const configJob = getConfig( {
    cwd: opts.cwd,
    configFiles: opts.configFiles ? flags.config || opts.configFiles : undefined,
    configKey: opts.configKey,
    failIfNotFound: !!flags.config
  } );

  if ( opts.notifier ) {
    notify( pkg, opts.notifier === true ? {} : opts.notifier );
  }

  const { config, location: configLocation } = await configJob;

  const flagsCopy = { ...flags };
  if ( opts.configFiles ) {
    delete flagsCopy.config;
  }
  delete flagsCopy.help;
  delete flagsCopy.version;

  flags[ "--" ] = argsInfo[ "--" ];

  const options = schema.validateAndMerge( {}, flagsCopy, config );

  return {
    command: commandSet.value,
    input,
    options,
    flags,
    config,
    configLocation,
    pkg,
    generateHelp,
    showHelp,
    showVersion
  };
}

export { cli, getConfig };
export default cli;
