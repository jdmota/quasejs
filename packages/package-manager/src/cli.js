// @flow
/* eslint-disable no-console */

import installer from "../src/installer";
import check from "../src/check";
import init from "../src/init";
import { read as readPkg, write as writePkg } from "../src/pkg";

const path = require( "path" );

function showDone() {
  console.log( `\nDone!\n` );
}

function showError( e: Error ) {
  console.error( `\n${e.message}\n` );
  process.exitCode = 1;
}

export function run( input: string[], flags: Object ) {

  const command = input[ 0 ] || "install";
  const folder = flags.folder ? path.resolve( flags.folder ) : process.cwd();

  switch ( command ) {

    case "install":
      installer( folder, flags ).then( showDone, showError );
      break;

    case "upgrade":
      flags.update = true;
      installer( folder, flags ).then( showDone, showError );
      break;

    case "normalize-pkg": {
      readPkg( folder ).then( pkg => writePkg( folder, pkg ) ).then( showDone, showError );
      break;
    }

    case "check":
      check( folder ).then( showDone, showError );
      break;

    case "init":
      init( folder, input[ 1 ] ).then( showDone, showError );
      break;

    default:
      showError( new Error( `Unknown ${command} command.` ) );
  }

}
