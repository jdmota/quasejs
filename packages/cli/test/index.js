import cli from "../src";

const { t } = require( "@quase/config" );

/* eslint-disable no-console */

describe( "cli", () => {
  it( "basic", async() => {

    process.stdout.isTTY = true;

    console.error = jest.fn();

    const { input, options, pkg, help, showHelp, config } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [ "foo", "--foo-bar", "--number", "10", "-u", "cat", "--boolean4=false", "--boolean5=true", "--", "unicorn", "10" ],
      help: `
        Usage
          foo <input>
      `,
      schema: {
        "--": { type: "array" },
        number: { default: 0 },
        unicorn: { alias: "u", optional: true },
        meow: { default: "dog" },
        boolean: { default: true },
        boolean2: { type: "boolean", default: true },
        boolean3: { type: "boolean" },
        boolean4: { type: "boolean" },
        boolean5: { type: "boolean" }
      },
      inferType: true,
      configFiles: "quase-cli-config.js",
      notifier: {
        options: {
          updateCheckInterval: 0
        },
        notify: {
          defer: false
        }
      }
    } );

    expect( input[ 0 ] ).toBe( "foo" );
    expect( options.fooBar ).toBe( true );
    expect( options.boolean ).toBe( true );
    expect( options.boolean2 ).toBe( true );
    expect( options.boolean3 ).toBe( false );
    expect( options.boolean4 ).toBe( false );
    expect( options.boolean5 ).toBe( true );
    expect( options.number ).toBe( 10 );
    expect( options.meow ).toBe( "dog" );
    expect( options.unicorn ).toBe( "cat" );
    expect( options[ "--" ] ).toEqual( [ "unicorn", "10" ] );
    expect( pkg.name ).toBe( "@quase/eslint-config-quase" );
    expect( pkg.version ).toBe( "0.0.1" );
    expect( help ).toMatchSnapshot();
    expect( typeof showHelp ).toBe( "function" );
    expect( config.iAmTheConfigFile ).toBe( "yes" );
    expect( console.error.mock.calls ).toMatchSnapshot();

  } );

  it( "defaults", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [ "--bar.prop2=1" ],
      schema: {
        foo: {
          type: "boolean"
        },
        bar: {
          type: t.object( {
            prop1: {
              default: 0
            },
            prop2: {
              default: 0
            }
          } )
        },
        baz: {
          type: t.object( {
            prop1: {
              default: 0
            },
            prop2: {
              default: 0
            }
          } ),
          default: {}
        }
      },
      notifier: false
    } );

    expect( options ).toEqual( {
      foo: false,
      bar: {
        prop1: 0,
        prop2: 1
      },
      baz: {}
    } );

  } );

  it( "generate help", async() => {

    const { help } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [],
      usage: "$ bin something",
      schema: {
        number: { default: 0, alias: [ "n", "n2" ], description: "number description" },
        unicorn: { alias: "u", optional: true, description: "unicorn description" },
        meow: { type: "string", default: "dog" },
        boolean: { type: "boolean", default: true, description: "boolean description" },
        object: {
          type: t.object( {
            prop: {
              type: "string",
              optional: true
            }
          } )
        },
        tuple: {
          type: t.tuple( [] ),
          description: "tuple"
        },
        array: {
          type: t.array( "string" ),
          description: "array of strings"
        },
        union: {
          type: t.union( [
            "string",
            "number"
          ] ),
          description: "union",
          optional: true
        },
        choices1: {
          choices: [ 0, 1, 2 ],
          description: "choices1",
          default: 0
        },
        choices2: {
          type: t.union( [ 0, 1, 2 ].map( t.value ) ),
          description: "choices2",
          default: 0
        },
        camelCase: {
          type: "string",
          description: "description",
          optional: true
        },
        noDescription: {
          description: "",
          optional: true
        }
      },
      notifier: false
    } );

    expect( help ).toMatchSnapshot();

  } );

  it( "inferType=true", async() => {

    const { options, input } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "--prop=10", "10", "abc" ],
      inferType: true,
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      prop: 10
    } );
    expect( input ).toEqual( [ 10, "abc" ] );

  } );

  it( "inferType=false", async() => {

    const { options, input } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "--prop=10", "10", "abc" ],
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      prop: 10
    } );
    expect( input ).toEqual( [ "10", "abc" ] );

  } );

  it( "alias", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "-a", "-b" ],
      schema: {
        aaa: {
          type: "boolean",
          alias: "a"
        },
        foo: {
          optional: true,
          argType: t.object( {
            bar: {
              type: "boolean",
              alias: "b"
            }
          } )
        }
      },
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      aaa: true,
      foo: {
        bar: true
      }
    } );

  } );

  it( "dot notation", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "--obj.foo=10", "--obj.bar.baz=dog", "--obj.bar.camelCase=cat" ],
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      obj: {
        foo: 10,
        bar: {
          baz: "dog",
          camelCase: "cat"
        }
      }
    } );

  } );

  it( "coerce", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "--obj={\"foo\":10}" ],
      schema: {
        obj: {
          type: "object",
          coerce( value ) {
            return {
              value: JSON.parse( value )
            };
          }
        }
      },
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      obj: {
        value: {
          foo: 10
        }
      }
    } );

  } );

  it( "count", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "-vvv" ],
      schema: {
        v: {
          type: "number",
          argType: "count",
          optional: true
        }
      },
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      v: 3
    } );

  } );

  it( "narg", async() => {

    const { options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "-a", "foo", "-b", "bar", "baz" ],
      schema: {
        aaa: {
          optional: true,
          narg: 1,
          alias: "a"
        },
        foo: {
          optional: true,
          argType: t.object( {
            bar: {
              narg: 2,
              alias: "b"
            }
          } )
        }
      },
      help: "",
      notifier: false
    } );

    expect( options ).toEqual( {
      aaa: "foo",
      foo: {
        bar: [ "bar", "baz" ]
      }
    } );

  } );

  it( "narg error", async() => {

    await expect(
      cli( {
        validate: false,
        cwd: __dirname,
        pkg: {
          name: "@quase/eslint-config-quase",
          version: "0.0.1"
        },
        argv: [ "-a", "foo", "--foo.bar", "bar", "baz" ],
        schema: {
          aaa: {
            narg: 1,
            alias: "a"
          },
          foo: {
            argType: t.object( {
              bar: {
                narg: 5
              }
            } )
          }
        },
        help: "",
        notifier: false
      } )
    ).rejects.toThrow( "Not enough arguments following: foo.bar" );

  } );

  it( "config", async() => {

    const { config, configLocation } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "--config=quase-cli-config-2.js" ],
      help: "",
      configFiles: "quase-cli-config.js",
      notifier: false
    } );

    expect( config.iAmTheConfigFile2 ).toBe( "yes" );
    expect( typeof configLocation ).toBe( "string" );

  } );

  it( "no config", async() => {

    const { config, configLocation } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [],
      help: "",
      configFiles: "non-existent-file.js",
      notifier: false
    } );

    expect( config ).toBe( undefined );
    expect( configLocation ).toBe( undefined );

  } );

  it( "config key", async() => {

    const { config, configLocation } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      argv: [ "" ],
      help: "",
      configFiles: "non-existent-file.js",
      configKey: "my-key",
      notifier: false
    } );

    expect( config.configFromPkg ).toBe( "yes" );
    expect( configLocation ).toBe( "pkg" );

  } );

  it( "config file has priority", async() => {

    const { options, configLocation } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1"
      },
      schema: {
        iAmTheConfigFile2: {
          type: "string",
          default: "no"
        }
      },
      argv: [ "--config=quase-cli-config-2.js" ],
      help: "",
      configFiles: "quase-cli-config.js",
      configKey: "my-key",
      notifier: false
    } );

    expect( options.iAmTheConfigFile2 ).toBe( "yes" );
    expect( options.configFromPkg ).toBe( undefined );
    expect( typeof configLocation ).toBe( "string" );
    expect( configLocation ).not.toBe( "pkg" );

  } );

  it( "asked for config that does not exist", async() => {

    await expect(
      cli( {
        validate: false,
        cwd: __dirname,
        pkg: {
          name: "@quase/eslint-config-quase",
          version: "0.0.1"
        },
        argv: [ "--config=non-existante-file.js" ],
        help: "",
        configFiles: [ "quase-cli-config.js" ],
        notifier: false
      } )
    ).rejects.toThrow( /^Config file was not found/ );

  } );

  it( "subcommands", async() => {

    const { command, input, options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      commands: {
        fooCommand: {
          schema: {
            foo: {
              type: "boolean"
            }
          }
        }
      },
      argv: [ "foo-command", "--foo" ],
      notifier: false
    } );

    expect( input ).toEqual( [] );
    expect( command ).toBe( "fooCommand" );
    expect( options.foo ).toBe( true );

  } );

  it( "invalid subcommand", async() => {

    await expect(
      cli( {
        cwd: __dirname,
        pkg: {
          name: "@quase/eslint-config-quase",
          version: "0.0.1",
          description: "Description"
        },
        commands: {
          fooCommand: {
            schema: {
              foo: {
                type: "boolean"
              }
            }
          }
        },
        argv: [ "bar-command", "--foo" ],
        notifier: false
      } )
    ).rejects.toThrow( /^"barCommand" is not a supported command$/ );

  } );

  it( "undefined subcommand", async() => {

    await expect(
      cli( {
        cwd: __dirname,
        pkg: {
          name: "@quase/eslint-config-quase",
          version: "0.0.1",
          description: "Description"
        },
        commands: {
          fooCommand: {
            schema: {
              foo: {
                type: "boolean"
              }
            }
          }
        },
        argv: [ "--foo" ],
        notifier: false
      } )
    ).rejects.toThrow( /^undefined is not a supported command$/ );

  } );

  it( "default subcommand", async() => {

    const { command, input, options } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      defaultCommand: "fooCommand",
      commands: {
        fooCommand: {
          schema: {
            foo: {
              type: "boolean"
            }
          }
        }
      },
      argv: [ "--foo" ],
      notifier: false
    } );

    expect( input ).toEqual( [] );
    expect( command ).toBe( "fooCommand" );
    expect( options.foo ).toBe( true );

  } );

  it( "generate help with subcommands", async() => {

    const { help } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [ "--help" ],
      usage: "$ bin <command> [options]",
      commands: {
        fooCommand: {
          description: "foo command description",
          schema: {
            number: { default: 0, alias: [ "n", "n2" ], description: "number description" }
          }
        },
        noDescription: {
          description: ""
        }
      },
      schema: {
        unicorn: { alias: "u", optional: true, description: "unicorn description" }
      },
      autoHelp: false,
      notifier: false
    } );

    expect( help ).toMatchSnapshot();

  } );

  it( "generate help with subcommands including options of default command", async() => {

    const { help } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [ "--help" ],
      defaultCommand: "fooCommand",
      usage: "$ bin <command> [options]",
      commands: {
        fooCommand: {
          description: "foo command description",
          schema: {
            number: { default: 0, alias: [ "n", "n2" ], description: "number description" }
          }
        }
      },
      schema: {
        unicorn: { alias: "u", optional: true, description: "unicorn description" }
      },
      autoHelp: false,
      notifier: false
    } );

    expect( help ).toMatchSnapshot();

  } );

  it( "generate help for subcommand", async() => {

    const { help } = await cli( {
      validate: false,
      cwd: __dirname,
      pkg: {
        name: "@quase/eslint-config-quase",
        version: "0.0.1",
        description: "Description"
      },
      argv: [ "fooCommand" ],
      usage: "$ bin <command> [options]",
      commands: {
        fooCommand: {
          description: "foo command description",
          schema: {
            number: { default: 0, alias: [ "n", "n2" ], description: "number description" },
            unicorn: { alias: "u", optional: true, description: "unicorn description" },
            meow: { type: "string", default: "dog" },
            boolean: { type: "boolean", default: true, description: "boolean description" },
          }
        }
      },
      schema: {},
      notifier: false
    } );

    expect( help ).toMatchSnapshot();

  } );

} );
