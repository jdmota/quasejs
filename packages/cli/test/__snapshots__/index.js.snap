// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli accept help for subcommand 1`] = `
"
  Description

  Custom subcommand help
"
`;

exports[`cli basic 1`] = `
"
  Description

  Usage
    foo <input>
      
"
`;

exports[`cli basic 2`] = `
Array [
  Array [
    "
                                                         
                                                         
             Update available [2m0.0.1[22m[0m â†’ [0m[32m0.2.0[39m              
    Run [36myarn add @quase/eslint-config-quase[39m to update    
                                                         
                                                         ",
  ],
]
`;

exports[`cli generate help 1`] = `
"
  Description

  Usage: $ bin something

  Options:
    --number, -n, -n2 number description (default: 0)
    --unicorn, -u     unicorn description
    --meow            (default: \\"dog\\") [string]
    --boolean         boolean description (default: false) [boolean]
    --boolean-no      boolean no description (default: true) [boolean]
    --object          [object]
    --object.prop     [string]
    --tuple           tuple [array]
    --array           array of strings [array]
    --union           union [string | number | array]
    --union.0         [string]
    --choices1        choices1 (default: 0) [0 | 1 | 2]
    --choices2        choices2 (default: 0) [0 | 1 | 2]
    --camel-case      description [string]
    --no-description 
"
`;

exports[`cli generate help for subcommand 1`] = `
"
  Description

  Usage: $ bin fooCommand [options]

  foo command description

  Options:
    --number, -n, -n2 number description (default: 0)
    --unicorn, -u     unicorn description
    --boolean         boolean description (default: true) [boolean]
"
`;

exports[`cli generate help with subcommands 1`] = `
"
  Description

  Usage: $ bin <command> [options]

  Commands:
    foo-command    foo command description
    no-description

  Options:
    --unicorn, -u unicorn description
"
`;

exports[`cli generate help with subcommands including options of default command 1`] = `
"
  Description

  Usage: $ bin <command> [options]

  Commands:
    foo-command foo command description [default]

  Options:
    --number, -n, -n2 number description (default: 0)
"
`;
